domains/pallets/auto-id/src/lib.rs

```diff
@@ -0,0 +1,407 @@
+                        issuer_auto_id.certificate.is_valid_at(current_time),
+                        Error::<T>::ExpiredCertificate
+                    );
+
+                    ensure!(
+                        !issuer_auto_id.certificate.is_revoked(),
+                        Error::<T>::CertificateRevoked
+                    );
+
+                    let tbs_certificate = decode_tbs_certificate(certificate.clone())
+                        .ok_or(Error::<T>::InvalidCertificate)?;
+
+                    let req = SignatureVerificationRequest {
+                        public_key_info: issuer_public_key_info,
+                        signature_algorithm,
+                        data: certificate.0.clone(),
+                        signature,
+                    };
+                    verify_signature(req).ok_or(Error::<T>::InvalidSignature)?;
+                    ensure!(
+                        tbs_certificate.validity.is_valid_at(current_time),
+                        Error::<T>::ExpiredCertificate
+                    );
+
+                    issuer_auto_id
+                        .certificate
+                        .issue_certificate_serial::<T>(tbs_certificate.serial)?;
+
+                    AutoIds::<T>::insert(issuer_id, issuer_auto_id);
+
+                    Certificate::X509(X509Certificate {
+                        issuer_id: Some(issuer_id),
+                        serial: tbs_certificate.serial,
+                        subject: tbs_certificate.subject,
+                        subject_public_key_info: tbs_certificate.subject_public_key_info,
+                        validity: tbs_certificate.validity,
+                        raw: certificate,
+                        issued_serials: BTreeSet::from([tbs_certificate.serial]),
+                        revoked: false,
+                    })
+                }
+            },
+        };
+
+        let auto_id_identifier = NextAutoIdIdentifier::<T>::get();
+        let next_auto_id_identifier = auto_id_identifier
+            .checked_add(Identifier::one())
+            .ok_or(Error::<T>::IdentifierOverflow)?;
+        NextAutoIdIdentifier::<T>::put(next_auto_id_identifier);
+
+        let auto_id = AutoId {
+            identifier: auto_id_identifier,
+            certificate,
+        };
+
+        AutoIds::<T>::insert(auto_id_identifier, auto_id);
+
+        Self::deposit_event(Event::<T>::NewAutoIdRegistered(auto_id_identifier));
+        Ok(())
+    }
+
+    fn do_verify_signature(auto_id: &AutoId, signature: Signature) -> DispatchResult {
+        let Signature {
+            signature_algorithm,
+            value: signature,
+        } = signature;
+        let req = SignatureVerificationRequest {
+            public_key_info: auto_id.certificate.subject_public_key_info(),
+            signature_algorithm,
+            // uses auto_id identifier as the message to sign
+            data: auto_id.identifier.encode(),
+            signature,
+        };
```

Comment:
This seems problematic, since the signing data is always the same the signature is always valid, so the attacker can reuse the signature from previous blocks to construct arbitrary request.

Reply:
Good catch!
We would need some extra random data for this block for users to sign along with AutoId, we can previous block hash. UX issue with this, whichever data we pick, transaction is valid only for one block. Users need to re-sign if the transaction does not make it into the block.
Thoughts @jfrank-summit

Reply:
>We would need some extra random data for this block for users to sign along with AutoId, we can previous block hash

This may not be enough as the attacker can listen to the tx pool and act something similar to MEV to front run the normal request. Ideally, we should sign the register/revoke/deactivate request itself.

Reply:
The problem still exists in that case as well since there is no specific request data for these actions.
I would need to think more this

Reply:
Actually, we can solve this with a nonce and request specific data. So even if the attacker does MEV, this signature will only be valid for specific request and specific nonce. They will end up paying for it instead of user.

Reply:
What does the user flow look like for this? Will we iterate nonces or track used random nonces?

Reply:
every certificate will have a signing nonce like we have a nonce for the account. Along with the nonce, we also will have the request specific pre-defined data. So the final data to be signed would look like this
```
Action enum {
	RevokeCert
	DeactivateAutoId
	...
}

SigningData struct {
	ID
	Nonce
	Action
}
```

This ensures, the signature is valid for this (nonce. action) and cannot be re-used.
This does not affect the X509 certificate signing. Just the actions done by the subject of the certificate